# Project Title

> - Create a title for the project. This can be fairly long, and should provide enough detail that someone can determine what this project is all about.
> - Sometimes it's helpful to include the UMGC project name in the title.

## Introduction

> - Include a short paragraph that briefly describes the basic background, questions, hypothesis, rationale, and goals for the analysis.
> - A second short paragraph should brefily outline the prior methods of the project and how the data were generated. This does not need to be detailed, but specific enough that someone new to the project can understand how the data were generated. Detailed descriptons of prior methods might be included in metadata files (e.g. grants, papers, abstracts) or the further described in your report.

## Project details

> - Establish a project name
>   - Project repos on GitHub and directory names on MSI can be different
>   - For the sake of uniform organization, create GitHub repo names in this format: `{group_id}_{user_id}_{unique_info}_{YYYYMMDD}`
>   - The directory name on MSI, should be created at this location, with this format: `/projects/standard/{group_id}/shared/ris/{user_id}/{unique_info}_{YYYYMMDD}`
> - Include the date for the start of the project (including day) because sometimes projects fall within or between months.
> - If you inherit a project (or collaborate on a project), each MSI analyst should create their own project dir. This will keep results and code separated. Linking to files in the related outputs, is ok (encouraged!).

The github.com/umn-msi-lmnp repo name is (for example, Todd working on a Mike Farrar project):

    farrarm_knut0297_wt_vs_gene_ko_scrnaseq_20250204

The directory name on MSI is:

    wt_vs_gene_ko_scrnaseq_20250204

## Contacts

> - List the persons involved in the project, their roles, and contact details.

### Contact roles

- NAME: PI of the project and MSI group.
- NAME: MSI bioinformatics analyst. Created this readme and started the project.
- NAME: Graduate student in PI lab.
- NAME: UMGC collaborator on custom library prep.

### Contact details

> - An quick method is to copy/paste email signatures.
> - Remember, standard Markdown format requires two spaces at the end of a line to signal a newline.

## Files

> - Show where the input project files are located and whether they were copied or stored in some other location.
> - Show where the analyzed project files can be found. Are they located on MSI, Google Drive, Box, Tier2 (ceph), etc.
> - List the project folder structure and where different types of information can be found.
> - Alternatively, if the project was inherited from another analyst, include the location of those raw and/or processed files.

### Raw data

- Raw sequencing data generated by UMGC were deposited at the location below (and will be automatically deleted after 1 year).

  ```
  /projects/standard/GROUP/data_delivery/umgc/RUN/PROJECT
  ```

- Raw sequencing data (above) were copied to MSI Tier2 (ceph) for long term storage. These data can be recovered if MSI hard drives fail, but not if MSI is destroyed. The data are stored in the bucket below.

  ```
  s3://data-delivery-GROUP
  s3://GROUP-data-archive
  ```

- Raw sequencing data (above) were copied to MSI Tier1 (panfs) for active usage. These data can be recovered if MSI hard drives fail or if MSI is destroyed because the data are stored in a special directory that is automatically copied to an offsite location.

  ```
  /projects/standard/GROUP/shared/disaster_recovery/ris/data_delivery/umgc/RUN/PROJECT
  ```

### Analyzed data

> - Provide a path to the project folder in the format `/projects/standard/GROUP/shared/ris/USER/PROJECT_NAME`
> - The path below sometimes might incldue a `ris` or `riss` dir. If a group does not already have a `ris/riss` dir, then skip it.
> - After establishing a private GitHub repo.
>   - Include a link to that repo below.
>   - You can add read or write access to the repo via GitHub teams (e.g. `lmnp_team` and `PI_team`).
>   - When creating the repo on GitHub, prefix the PI group and your username to the project name in this format `{GROUP}_{USER}_{PROJECT_NAME}` (e.g. `farrarm_knut0297_wt_vs_gene_ko_scrnaseq_20250204`).

The project directory (below) can be found on MSI Tier 1 storage. It contains all of the input files (or symlinks to input files), source code, output files, and metadata files.

    /projects/standard/GROUP/shared/ris/USER/wt_vs_gene_ko_scrnaseq_20250204

Inside the project directory above, various sub-directories can be found and are described below:

| Directory        | Description                                                                                                                                                                                                                                                                                                                      |
| ---------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| code             | All analysis and documentation code is saved to the code directory (dir).                                                                                                                                                                                                                                                        |
| code_out         | Output dirs that were generated by a script in the code dir. The name of a script corresponds to the name of the output dir (e.g. the `demo_r_analysis.slurm` script in the code dir will create a folder and results called `demo_r_analysis` in the code_out dir). Everything in this dir can be recreated by scripts in the code dir. |
| input            | This dir contains any files that were manually created or supplied from outside this project, and used by scripts in the code dir.                                                                                                                                                                                               |
| meta             | This metadata dir contains any files important to the project (e.g. Word docs, presentations, abstracts, grants, sequencing metadata, bed files, etc.), but are not used as input files during the analysis.                                                                                                                     |
| docs             | If present, this dir contains summarized project documentation or reports. This folder contains documentation manually generated (i.e. RMarkdown files, etc. would not be found here).                                                                                                                                           |
| software         | Contains build scripts and configuration files for creating reproducible software environments (conda and Apptainer).                                                                                                                                                                                                            |
| software_build   | Contains all installed software (conda environments, packages, containers). This directory is NOT tracked in git and is reproducibly built from scripts in software/.                                                                                                                                                             |
| .apptainer       | Contains Apptainer container build cache and temp files. This directory is NOT tracked in git.                                                                                                                                                                                                                                   |
| changelog.md     | A Markdown formatted file that contains a reverse chronological description of analysis items.                                                                                                                                                                                                                                   |
| readme.md        | A Markdown formatted file that contains a brief description of the project name and where to find the data. This file is the readme file displayed in the GitHub repo main page.                                                                                                                                                 |
| .git             | This dir contains everything related to git tracking (This is a hidden dir on most systems).                                                                                                                                                                                                                                     |
| .gitignore       | This file contains a list of files/dirs that should not be tracked by git (This is a hidden file on most systems).                                                                                                                                                                                                               |

Only the source code, select input files, and metadata files are being version controlled with git and synced with UMN Enterprise GitHub. Large input files and all output files and not stored in GitHub.

The repository can be found here and will require a UMN Internet ID/password for access:

[https://github.com/umn-msi-lmnp/wt_vs_gene_ko_scrnaseq_20250204](https://github.com/umn-msi-lmnp/wt_vs_gene_ko_scrnaseq_20250204)

## Software environments

This project uses self-contained, reproducible software environments. All software is installed locally within the project directory.

### Key features

- **Self-contained**: All software stays within this project, not in $HOME
- **Reproducible**: Environments are built from version-controlled YAML files
- **Portable**: Clone this project anywhere and rebuild identical environments
- **No conflicts**: Independent of system-wide or user-level conda installations

### Software locations within this project

All software is located in the `software_build/` directory (not tracked in git):

```
software_build/
  miniforge/
    envs/
      env1/          # R environment (R, tidyverse, ggplot2, DESeq2, Seurat)
      env2/          # Python environment (python, pandas, biopython, samtools)
    pkgs/            # All conda package cache (shared across environments)
  containers/
    picard.sif       # Picard Tools Apptainer container

.apptainer/
  cache/             # Apptainer container build cache
  tmp/               # Apptainer temporary files
```

### R package libraries

R packages from env1 are installed in:
```
software_build/miniforge/envs/env1/lib/R/library/
```

To check library paths in R:
```r
.libPaths()
```

### Python packages

**From env2** (Python/bioinformatics environment):
```
software_build/miniforge/envs/env2/lib/python3.X/site-packages/
```

**From env1** (via pip in R environment):
```
software_build/miniforge/envs/env1/lib/python3.X/site-packages/
```

env1 includes Python and pip for installing Python packages needed by R packages (e.g., `leidenalg` for `r-leiden`). Add pip packages to `software/environment1.yml`:

```yaml
dependencies:
  - r-leiden
  - python>=3.9
  - pip
  - pip:
      - leidenalg
```

### Dependencies

This project requires only one system-level dependency:
- **Apptainer** (for building and running containers)

Everything else (conda, mamba, R, Python, all packages) is installed locally within the project.

### Building environments

See `software/` directory for build scripts:
```bash
cd software
sbatch 010_build_minforge.slurm           # Build local conda installer
sbatch 011_build_conda_environment1.slurm # Build R environment (env1)
sbatch 012_build_conda_environment2.slurm # Build Python environment (env2)
sbatch 021_build_apptainer_picard.slurm   # Build Picard container
```

All environments can be rebuilt at any time using these scripts. The `software_build/` directory can be deleted and recreated from scratch.

## Report

> - The project report should include summary analysis and figures (and potentially code, if using Rmarkdown-like docs). The report could be a Word doc, Google Doc, text file, etc.
> - Show where and how the report can be accessed. This does not need to be a new directory (i.e. `docs` as suggested above), but can be somewhere in the outputs.
> - For example, if you generate markdown files as outputs from your scripts, point the user to where those can be found.
> - Or, if you summarized your results in a Google Doc, Powerpoint, etc., provide a link or path to these files.
> - If you have files related to a manuscript, methods, etc. you can provide ways to access them here, including GEO accession numbers, etc.

## Changelog

> - All notable changes to this project will be documented. See [https://keepachangelog.com](https://keepachangelog.com).
> - A changelog is a file which contains a curated, chronologically ordered list of notable changes for each version of a project.
> - To make it easier for users and contributors to see precisely what notable changes have been made between each update.
> - Write this for people: for your collaborators and your future self.
> - Ask the collaborators write a summary email after meetings so we can maintain priorities.

A `changelog.md` file is maintained at the root of this project that briefly lists items that were new, changed, removed, fixed, or future to-dos.
